README.md
Tour (Roundcube Webmail Plugin)
==========
Plugin for adding a tour to the site.

Translation
-----------
You can help me to translate plugin [here](https://www.transifex.com/san4op/tour-roundcube-webmail-plugin/dashboard/).

Donate
------
You can make a donation [here](http://yasobe.ru/na/roundcube_tour), this will help motivate me to continue my work.


CHANGELOG
CHANGELOG
=========

RELEASE 1.2
-----------
* Added a button to manually run tour
* Small internal changes

RELEASE 1.1
-----------
* Support of 'managesieve' & 'password' plugins
* Improved checking of enabled plugins
* Improved compatabile with Roundcube < 1.1.0
* Update Intro.js to 2.7.0

RELEASE 1.0
-----------
* First release


tour.js
/**
 * Roundcube Plugin Tour
 * Plugin for adding a tour to the site.
 *
 * @version 1.2
 * @author Alexander Pushkin <san4op@icloud.com>
 * @copyright Copyright (c) 2017, Alexander Pushkin
 * @link https://github.com/san4op/roundcube_tour
 * @license GNU General Public License, version 3
 */

if (window.rcmail) {
	function tour(manually) {
		var intros = [];

		if (rcmail.task == "mail") {
			if (rcmail.env.action == "") {
	
				// welcome
				if (rcmail.env.tour.welcome == true)
					intros.push({
						element: '#rcmbtn105',
						intro: rcmail.gettext('welcome', 'tour'),
						position: 'bottom-left-aligned'
					});

				// support
				if (rcmail.env.tour.support == true)
					intros.push({
						element: '#supportlink',
						intro: rcmail.gettext('support', 'tour'),
						position: 'bottom-left-aligned'
					});
	
				// taskbar
				if (rcmail.env.tour.taskbar == true)
					intros.push({
						element: '#taskbar',
						intro: rcmail.gettext('taskbar', 'tour'),
						position: 'bottom-right-aligned'
					});
	
				// taskbar buttons
				if (rcmail.env.tour.taskbar_buttons)
					[
						"mail",
						"addressbook",
						"cloud",
						"calendar",
						"files",
						"notes",
						"tasklist",
						"settings"
					].forEach(function(elm,inx,arr) {
						if (typeof rcmail.env.tour.taskbar_buttons[elm] != "undefined" && rcmail.env.tour.taskbar_buttons[elm] == true)
							intros.push({
								element: '#taskbar .button-' + elm,
								intro: rcmail.gettext('taskbar_' + elm, 'tour'),
								position: 'bottom-right-aligned'
							});
					});
	
				// toolbar
				if (rcmail.env.tour.toolbar == true)
					intros.push({
						element: '#messagetoolbar',
						intro: rcmail.gettext('toolbar', 'tour'),
						position: 'bottom-right-aligned'
					});
	
				// toolbar buttons
				if (rcmail.env.tour.toolbar_buttons)
					[
						"archive",
						"junk"
					].forEach(function(elm,inx,arr) {
						if (typeof rcmail.env.tour.toolbar_buttons[elm] != "undefined" && rcmail.env.tour.toolbar_buttons[elm] == true)
							switch(elm) {
								case "junk":
									intros.push({
										element: '#messagetoolbar .button.junk, #messagetoolbar .button.markasjunk2',
										intro: rcmail.gettext('toolbar_' + elm, 'tour'),
										position: 'bottom'
									});
									break;
	
								default:
									intros.push({
										element: '#messagetoolbar .button.' + elm,
										intro: rcmail.gettext('toolbar_' + elm, 'tour'),
										position: 'bottom'
									});
									break;
							}
					});
	
				// folders
				if (rcmail.env.tour.folders == true)
					intros.push({
						element: (rcmail.env.skin == "classic" ? "#mailboxlist-container ul" : "#folderlist-content ul"),
						intro: rcmail.gettext('folders', 'tour'),
						position: 'right'
					});
	
				// quota
				if (rcmail.env.tour.quota == true)
					intros.push({
						element: (rcmail.env.skin == "classic" ? "#quota" : "#quotadisplay"),
						intro: rcmail.gettext('quota', 'tour'),
						position: 'top'
					});
	
				// messageslist
				if (rcmail.env.tour.messages_view == true)
					intros.push({
						element: (rcmail.env.layout == 'widescreen' ? '#listmenulink' : '.messagelist.fixedcopy #rcmthreads #listmenulink'),
						intro: rcmail.gettext('messages_view', 'tour'),
						position: 'bottom'
					});
				if (rcmail.env.tour.messages_threads == true && rcmail.env.skin != "classic")
					intros.push({
						element: '#listcontrols',
						intro: rcmail.gettext('messages_threads', 'tour'),
						position: 'top'
					});

				// tags
				if (rcmail.env.tour.taglist == true)
					intros.push({
						element: '#taglist',
						intro: rcmail.gettext('taglist', 'tour'),
						position: 'bottom-right-aligned'
					});
	
			} // end action ""
		} // end task "mail"

		else if (rcmail.task == "settings") {
			if (rcmail.env.action == "") {
	
				// settings
				if (rcmail.env.tour.settings)
					[
						"preferences",
						"folders",
						"identities",
						"responses",
						"pluginmanagesieve",
						"pluginmanagesievevacation",
						"pluginpassword"
					].forEach(function(elm,inx,arr) {
						if (typeof rcmail.env.tour.settings[elm] != "undefined" && rcmail.env.tour.settings[elm] == true)
							intros.push({
								element: '#settingstab' + elm,
								intro: rcmail.gettext('settings_' + elm, 'tour'),
								position: 'right'
							});
					});
	
			} // end action ""
		} // end task "settings"

		if (intros.length > 0) {
			var intro = introJs();
			var intro_complete = 0;
	
			intro.setOptions({
				nextLabel: rcmail.gettext('next', 'tour'),
				prevLabel: rcmail.gettext('prev', 'tour'),
				skipLabel: rcmail.gettext('skip', 'tour'),
				doneLabel: rcmail.gettext('done', 'tour'),
				hidePrev: true,
				hideNext: true,
				showStepNumbers: false,
				showBullets: false,
				disableInteraction: true,
				exitOnOverlayClick: false,
				tooltipPosition: "bottom"
			});
			intro.addSteps(intros);
	
			if (typeof manually == "undefined" || manually != true) {
				intro.oncomplete(function() {
					intro_complete = 1;
					rcmail.http_post('plugin.tour_complete', {task: rcmail.task, action: rcmail.env.action, complete: 2});
				});
				intro.onexit(function() {
					if(typeof intro_complete == "undefined" || intro_complete == 0) {
						rcmail.http_post('plugin.tour_complete', {task: rcmail.task, action: rcmail.env.action, complete: 1});
					}
				});
			}
	
			intro.start();
		}
	}

	rcmail.addEventListener('init', function(evt) {
		if (typeof rcmail.env.tour_run != "undefined" && rcmail.env.tour_run == 1) {
			tour();
		}
	});
}


tour.php
<?php
/**
 * Roundcube Plugin Tour
 * Plugin for adding a tour to the site.
 *
 * @version 1.2
 * @author Alexander Pushkin <san4op@icloud.com>
 * @copyright Copyright (c) 2017, Alexander Pushkin
 * @link https://github.com/san4op/roundcube_tour
 * @license GNU General Public License, version 3
 */

class tour extends rcube_plugin
{
	public $task = 'login|mail|settings';
	public $noframe = true;
	private $rc;
	private $prefs;

	function init()
	{
		$this->rc = rcube::get_instance();

		// show don't completed tour again after login
		if ($this->rc->task == 'login') {
			$this->add_hook('login_after', array($this, 'user_login'));
			return;
		}

		// get status of tour
		$this->prefs = $this->rc->user->get_prefs();
		$this->prefs = (isset($this->prefs['tour']) ? $this->prefs['tour'] : array());

		// load config
		$this->load_config();
		$introjs_theme = $this->rc->config->get('tour_introjs_theme', false);

		// load localization
		$this->add_texts('localization/', true);

		// include styles
		$this->include_stylesheet('lib/intro.js/introjs.css');
		if ($introjs_theme) {
			$this->include_stylesheet('lib/intro.js/themes/introjs-'.$introjs_theme.'.css');
		}
		$this->include_stylesheet($this->local_skin_path() . '/tour.css');

		// include scripts
		$this->include_script('lib/intro.js/intro.js');
		$this->include_script('tour.js');

		// add button
		$this->add_button(
			array(
				'name' => 'tour',
				'type' => 'link',
				'label' => 'tour.button',
				'onclick' => 'if(tour)tour(true);return false',
			),
			'topline-left'
		);

		// load config for buttons
		$taskbar_buttons = $this->rc->config->get('tour_taskbar_buttons', array('mail' => true, 'addressbook' => true, 'settings' => true));
		$toolbar_buttons = $this->rc->config->get('tour_toolbar_buttons', array());
		$settings_actions = $this->rc->config->get('tour_settings', array('preferences' => true, 'folders' => true, 'identities' => true, 'responses' => true));

		// check loaded plugins
		if (version_compare(RCMAIL_VERSION, '1.1') >= 0) {
			$plugins = $this->api->active_plugins;
		} else {
			$plugins = array_filter((array) $this->rc->config->get('plugins'));
		}

		if (!in_array('cloud_button', $plugins)) {
			$taskbar_buttons['cloud'] = false;
		}

		if (!in_array('calendar', $plugins)) {
			$taskbar_buttons['calendar'] = false;
		}

		if (!in_array('tasklist', $plugins)) {
			$taskbar_buttons['tasklist'] = false;
		}

		if (!in_array('kolab_notes', $plugins)) {
			$taskbar_buttons['notes'] = false;
		}

		if (!in_array('kolab_files', $plugins)) {
			$taskbar_buttons['files'] = false;
		}

		if (!in_array('archive', $plugins) || !$this->rc->config->get('archive_mbox')) {
			$toolbar_buttons['archive'] = false;
		}

		if (!in_array('markasjunk', $plugins) && !in_array('markasjunk2', $plugins)) {
			$toolbar_buttons['junk'] = false;
		}

		if (!in_array('managesieve', $plugins)) {
			$settings_actions['pluginmanagesieve'] = false;
			$settings_actions['pluginmanagesievevacation'] = false;
		} else {
			if (version_compare(RCMAIL_VERSION, '1.1') >= 0) {
				if (($managesieve = $this->api->get_plugin('managesieve')) instanceof managesieve) {
					if ($managesieve->load_config()) {
						$vacation_mode = (int) $this->rc->config->get('managesieve_vacation');
						if ($vacation_mode === 2) {
							$settings_actions['pluginmanagesieve'] = false;
						}
						if ($vacation_mode === 0) {
							$settings_actions['pluginmanagesievevacation'] = false;
						}
					} else {
						$settings_actions['pluginmanagesievevacation'] = false;
					}
				} else {
					$settings_actions['pluginmanagesieve'] = false;
					$settings_actions['pluginmanagesievevacation'] = false;
				}
				unset($managesieve);
			} else {
				$settings_actions['pluginmanagesieve'] = false;
				$settings_actions['pluginmanagesievevacation'] = false;
			}
		}

		if (!in_array('password', $plugins)) {
			$settings_actions['pluginpassword'] = false;
		}

		// make enviroment
		$this->api->output->set_env('tour', array(
			'welcome'			=> $this->rc->config->get('tour_welcome', true),
			'support'			=> $this->rc->config->get('tour_support', true),
			'taskbar'			=> $this->rc->config->get('tour_taskbar', true),
			'taskbar_buttons'	=> $taskbar_buttons,
			'toolbar'			=> $this->rc->config->get('tour_toolbar', true),
			'toolbar_buttons'	=> $toolbar_buttons,
			'folders'			=> $this->rc->config->get('tour_folders', true),
			'quota'				=> $this->rc->config->get('tour_quota', true),
			'taglist'			=> $this->rc->config->get('tour_taglist', true),
			'messages_view'		=> $this->rc->config->get('tour_messages_view', true),
			'messages_threads'	=> $this->rc->config->get('tour_messages_threads', true),
			'settings'			=> $settings_actions,
		));

		// not run again if tour already complete
		if (isset($this->prefs[$this->rc->task][$this->rc->action]) && $this->prefs[$this->rc->task][$this->rc->action] > 0) {
			return;
		}

		// register action to process complete of tour
		$this->register_action('plugin.tour_complete', array($this, 'complete'));

		// show tours
		if (!$this->rc->output->ajax_call) {
			$this->api->output->set_env('tour_run', 1);
		}
	}

	public function complete()
	{
		if (!$this->rc->output->ajax_call) {
			return;
		}

		$task = rcube_utils::get_input_value('task', rcube_utils::INPUT_POST);
		$action = rcube_utils::get_input_value('action', rcube_utils::INPUT_POST);
		$complete = (int) rcube_utils::get_input_value('complete', rcube_utils::INPUT_POST);

		if ($complete > 0) {
			if (!isset($this->prefs[$task])) {
				$this->prefs[$task] = array($action => $complete);
			}
			elseif (!isset($this->prefs[$task][$action]) || $this->prefs[$task][$action] <> 2) {
				$this->prefs[$task][$action] = $complete;
			}

			$this->rc->user->save_prefs(array('tour' => $this->prefs));
		}
	}

	public function user_login()
	{
		$prefs = $this->rc->user->get_prefs();
		$prefs = (isset($prefs['tour']) ? $prefs['tour'] : array());

		foreach ($prefs as $task => $actions) {
			foreach ($actions as $action => $complete) {
				if($complete == 1) {
					$prefs[$task][$action] = 0;
				}
			}
		}

		$this->rc->user->save_prefs(array('tour' => $prefs));
	}
}

?>
